Cargo.toml

```diff
@@ -16,7 +16,7 @@ members = [
     "ast",
     "analysis",
     "linker",
-    "asm_utils"
```

Comment:
Our eStark integration should be its own crate, similar to halo2

Reply:
This could be a separate PR but how about:
- one `backend_traits` crate with only the traits common to all backends
- one crate per backend which implements these traits
- one backend dispatcher crate which does the dynamic dispatching

Reply:
Many things comes to my mind:
* It is a good way to avoid circular dependency between `backend` crate and `halo2`, and still have almost everything halo2 implemented in `halo2` crate.
* If we are going to do it, leave estark where it is in this PR, and change it in that PR.
* I don't see a point in this separation into crates, and I prefer placing everything (including `halo2`) in `backend` crate.

Reply:
Also:
* I am thankful zokrates separated everything into crates, so it was easy to reuse some stuff inside polynomial-solver.

Reply:
The current way we have things is inconsistent:
- halo2 is a separate crate but some of it is in backend
- estark is entirely in backend

So it needs to be fixed either way.

> * It is a good way to avoid circular dependency between `backend` crate and `halo2`, and still have almost everything halo2 implemented in `halo2` crate.
I don't think there is a circular dependency if we define a backend trait and implement it in backend crates.
> * If we are going to do it, leave estark where it is in this PR, and change it in that PR.
Yes, I think this is fine for this PR.
> * I don't see a point in this separation into crates, and I prefer placing everything (including `halo2`) in `backend` crate.
Crates force us to have clean interfaces (fixing circular dependencies cleanly) and makes incremental builds faster. Here for example, cleanly separating the trait that a backend must implement, the backends, and the way we dispatch to a backend, currently dynamically.

Reply:
Ok. In another PR.

Reply:
Shouldn't the `estark` stuff at least be moved to its own crate for now? I just merge as is and do it together with the new split/refactoring?